corehq/apps/es/users.py

```diff
@@ -73,9 +75,51 @@ class ElasticUser(ElasticDocumentAdapter):
     def mapping(self):
         return get_adapter_mapping(self)
 
-    @classmethod
-    def from_python(cls, doc):
-        return from_dict_with_possible_id(doc)
+    def from_python(self, user):
+        """
+        :param user: an instance of ``CouchUser`` or a user dict
+        :raises ``TypeError`` user is none of the above types
+        """
+        from corehq.apps.users.models import CouchUser
+        if isinstance(user, CouchUser):
+            user_dict = user.to_json()
+        elif isinstance(user, dict):
+            user_dict = user
```

Comment:
Copying discussion context from [other comment](https://github.com/dimagi/commcare-hq/pull/32660#discussion_r1116686135):

> > what do you think of always passing in a dict that can be mutated, so the decision to copy or not can be pushed up to `from_python` which can make a more intelligent decision? For example, no copy is necessary when the thing being passed in here came from `user.to_json()`.
> 
> The idea crossed my mind when I was implementing it, but I held back from doing it because we would be forced to use `deepcopy` everytime. Only that would ensure that the dict passed is safe to mutate. If we give this responsibility to `from_dict` then we could judicially copy it as per our requirements.

Are you still intending to implement this, or did I miss the reason we don't want to push `copy` up to `from_python` now that `_from_dict` is private?

Reply:
I have a different thought process here-
We have two scenarios, first when we want `from_dict` the ability to alter the dicts that are being passed in. In that case we would have to deepcopy the dicts and pass it to `from_dict`
The second case is not allowing `from_dict` to mutate the dict which is happening right now. In this case `from_dict` either copies or deepcopies based on the requirements and never alters the dicts.

If we push `copy` to `from_python` it will create a scenario where we sometimes would be mutating the dict and sometime we will not be. I am not sure if it the right way to go.